hathor/indexes/memory_deps_index.py

```diff
@@ -0,0 +1,170 @@
+
+        # Needed txs (key: tx missing, value: requested by)
+        self._needed_txs_index: Dict[bytes, Tuple[int, bytes]] = {}
+
+    def update_validation(self, tx: bytes, validation: ValidationState) -> None:
+        if tx in self._rev_dep_index:
+            self._update_validation(tx, validation)
+
+    def _update_validation(self, tx: bytes, validation: ValidationState) -> None:
+        """Propagate the new validation state of a given dep."""
+        for cousin in self._rev_dep_index[tx].copy():
+            deps = self._dir_dep_index[cousin]
+            # XXX: this check serves to avoid calling is_ready() when nothing changed
+            if deps[tx] != validation:
+                deps[tx] = validation
+                if deps.is_ready():
+                    self._del_from_deps_index(cousin)
+                    self._txs_with_deps_ready.add(cousin)
+
+    def update(self, tx: BaseTransaction, partial: bool = True) -> None:
+        assert tx.hash is not None
+        assert tx.storage is not None
+        validation = get_validation_state(tx.storage, tx.hash)
+        if validation.is_fully_connected():
+            self._del_from_deps_index(tx.hash)
+        elif not partial:
+            raise ValueError('partial=False will only accept fully connected transactions')
+        else:
+            self._add_deps(tx)
+            self._add_needed(tx)
+
+    def _add_deps(self, tx: BaseTransaction) -> None:
+        assert tx.hash is not None
+        assert tx.storage is not None
+        # deps are immutable for a given hash
+        _deps = _DirDepValue((dep, get_validation_state(tx.storage, dep)) for dep in tx.get_all_dependencies())
+        # short circuit add directly to ready
+        if _deps.is_ready():
+            self._txs_with_deps_ready.add(tx.hash)
+            return
+        # add direct deps
+        if __debug__ and tx.hash in self._dir_dep_index:
+            # XXX: dependencies set must be immutable
+            assert self._dir_dep_index[tx.hash].keys() == _deps.keys()
+        self._dir_dep_index[tx.hash] = _deps
+        # add reverse dep
+        for rev_dep in _deps:
+            if rev_dep not in self._rev_dep_index:
+                self._rev_dep_index[rev_dep] = set()
+            self._rev_dep_index[rev_dep].add(tx.hash)
+
+    def _del_from_deps_index(self, tx: bytes) -> None:
+        _deps = self._dir_dep_index.pop(tx, _DirDepValue())
+        for rev_dep in _deps.keys():
+            rev_deps = self._rev_dep_index[rev_dep]
+            if tx in rev_deps:
+                rev_deps.remove(tx)
+            if not rev_deps:
+                del self._rev_dep_index[rev_dep]
+
+    def remove_ready_for_validation(self, tx: bytes) -> None:
+        """ Removes from ready for validation set.
+        """
+        self._txs_with_deps_ready.discard(tx)
+
+    def next_ready_for_validation(self, tx_storage: 'TransactionStorage', *, dry_run: bool = False) -> Iterator[bytes]:
```

Comment:
This code seems really confusing. It seems the caller might change `_txs_with_deps_ready`, so you have to copy it before yielding. In order to iterate over the new additions to the `_txs_with_deps_ready`, you need this while loop that updated `cur_ready` at the end. Is that right?

Reply:
This method is designed to allow being called only once and still yield new items that might be added because of external side effects (since the iterator will be consumed specifically while the transactions it yields are validated and as a side effect more transactions become ready for validation).

I agree that it is confusing, because there are some non-obvious caveats, if dry_run=False and you don't consume the whole iterator, the list of "transactions ready" will be cleared and we would not validate some transactions. Although the only place that uses this method consumes the iterator until the end.

These caveats can certainly be remove, although maybe with a small performance loss (currently it will sort the transactions everytime the list is exhausted so it should be something like `O(k.n.log(n))` and `k << n`, so it's more like `O(n.log(n))`, and the trivial fix for the caveat would just take the `min` (or `max`, not sure) of the list at each iteration, so it would be something like `O(m.n)` but since `m ~ n`, it's more like `O(n2)`, but this is really from the top of my head and lacking a lot of explanation, I may be missing something).

The `copy` is only for `dry_run=True`, which I think might not be in use anymore (I think I only used for testing). So it doesn't affect its real use. The `while` is just for allowing the iterator to be used until it is exhausted (otherwise the user of this function would basically need a part of the structure that is inside of this function.

Anyway, I think I remember it better. This function was refactored to simplify the caller, I think the `def method() -> Iterator[bytes]` does this well for the caller, but maybe we can drop the `dry_run` and make some refactors and remove the caveats for this implementation, even if at a slightly less efficient way, since we'll eventually want the rocksdb implementation to be the default (which may be easier to optimize).

Reply:
I understood how it works. We have other discussion threads covering potential issues. So I'll close this thread.